go-metadata/istio.go

```diff
@@ -0,0 +1,132 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package istiometadata
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"time"
+)
+
+// ConnectionMetadata returns the metadata about a connection
+type ConnectionMetadata struct {
+	Identity string `json:"identity"`
+}
+
+// metadataClient provides a default client for all metadata requests
+var metadataClient = http.Client{
+	Timeout: time.Second,
+}
+
+// metadataContextKey is the key the metadata handler will store connection metadata in
+var metadataContextKey = &contextKey{}
+
+type contextKey struct {
+	name string
+}
+
+func (k *contextKey) String() string { return "istio metadata context value" }
+
+func metadataServerURL() *url.URL {
+	u, err := url.Parse("http://169.254.169.111")
+	if err != nil {
+		panic(err.Error())
+	}
+	return u
+}
+
+func lookup(src, dst string) (*ConnectionMetadata, error) {
+	srcip, srcport, err := net.SplitHostPort(src)
+	if err != nil {
+		return nil, fmt.Errorf("invalid src: %v", err)
+	}
+	dstip, dstport, err := net.SplitHostPort(dst)
+	if err != nil {
+		return nil, fmt.Errorf("invalid dst: %v", err)
+	}
+
+	u := metadataServerURL()
+	params := url.Values{}
+	params.Add("srcip", srcip)
+	params.Add("srcport", srcport)
+	params.Add("dstip", dstip)
+	params.Add("dstport", dstport)
+	u.RawQuery = params.Encode()
+
+	resp, err := metadataClient.Get(u.String())
```

Comment:
I  would add a path to the request - we may add more MDS queries later. 

Also not sure of query param encoding is best - but we can start with it. 
Maybe we shouldn't split the host port - and let Rust take care of that or just use the unsplit form as key, less
garbage and processing ?